### PR TITLE
implement proposal for fetching paths and retaining structure of dirs and packages

### PR DESCRIPTION
**Pull request by madhur-tandon**
Closes https://github.com/pyscript/pyscript/issues/519 and https://github.com/pyscript/pyscript/issues/921

Proposal here: https://github.com/pyscript/pyscript/issues/519#issuecomment-1276276317

TODO:
- [x]  Update docs

### CODE REVIEW

File: docs/reference/elements/py-config.md
Comment (antocuni) [inline docs/reference/elements/py-config.md:80]: [suggestion] This example has the purpose of showing that you can use both `src` and inline content, it's probably better to use an easier-to-understand config settings, e.g. `packages` or `name`.

File: docs/reference/elements/py-config.md
Comment (antocuni) [inline docs/reference/elements/py-config.md:231]: I think it's fine for this PR, but we should have better docs to explain how this works.
@tedpatrick do we have a process to signal "please write better docs for this topic" so that our technical writers know about it?
E.g. by putting `TODO: bla bla bla` comments directly inside the docs or creating an issue?

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts:115]: why did yoy put this code into `utils.ts`?
It's specifically for `py-config`, it should go into `pyconfig.ts.`

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts:118]: [question] just my curiosity but: what is the advantage of using `forEach` instead of a normal `for (each_fetch_cfg of FetchConfig)`?

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts:122]: this is missing support for `to_file` as described here: https://github.com/pyscript/pyscript/issues/519#issuecomment-1302362956

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts:131]: I have no idea what this line does :). Sure, I could probably spend 5 minutes trying to understand it, OR you could write a nice function. If I understand correctly, there you want to do essentially the equivalent of `os.path.join`, don't you?
If so, I suggest to just write a small function `joinPaths` which does just this.
Bonus point: as soon as you have your own function, you can unit-test it very well, with all sort of corner cases.

Also: why do you have `filter(it => it !== "")`? Does it mean that if I have `files = ["a", "", "b"]`, the 2nd is silently ignored and it ends up fetching only 3 files?
If so, -1. It should be an error.

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts:145]: oh, and this is the 3rd time you copy&pasted the code. One more reason to make it a standalone function

File: pyscriptjs/src/utils.ts
Comment (antocuni) [inline pyscriptjs/src/utils.ts]: small note: if PR #909 gets merged before this one, this shoud become `throw UserError`.

**Review by antocuni — CHANGES_REQUESTED** [review]
Good job, apart a couple of notes.
But, it completely misses integration tests which should to to `tests/integration/test_py_config.py`.

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:122]: @antocuni I think you reviewed a previous commit. I already added support for `to_file`. Not sure why you cannot see it.

File: docs/reference/elements/py-config.md
Comment (madhur-tandon) [reply docs/reference/elements/py-config.md:231]: I guess you cannot see an entry for `to_file` here as well since I think you reviewed a previous commit.

File: docs/reference/elements/py-config.md
Comment (madhur-tandon) [reply docs/reference/elements/py-config.md:80]: done, using `packages`

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:115]: put it in `plugins/fetch.ts` as discussed on slack

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:145]: done

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:131]: done, added testing, the part about `filter(it => it !== "")` is needed since the behaviour of
`os.path.join('', '', '')` is simply `''` so we are replicating that in some sense.

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts:122]: resolving...

**Comment by madhur-tandon** [comment]
@antocuni For the integration tests, they were present in `test_01_basic.py`, I have moved it to `test_py_config.py`.
Let me know if you have ideas for more of them.

I am not sure if we should write tests using a custom URL (eg: `http://example.com/data.csv`) -- since:
- I don't know such example websites that host some dummy files
- Do we want tests that depend on network?

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts]: I don't know to be honest except that it is considered a standard practice I guess? If not, I can change. Let me know.

Comment (antocuni) [reply pyscriptjs/src/utils.ts]: I don't really know because I'm not an expert of JS/TS.
In all the rest of the code we use `for(... of ...)`, that's why I wondered whether here it was different

**Comment by antocuni** [comment]
> I am not sure if we should write tests using a custom URL (eg: `http://example.com/data.csv`) -- since:

if this case is already covered by unit tests, I think we are fine

File: pyscriptjs/tests/unit/utils.test.ts
Comment (antocuni) [inline pyscriptjs/tests/unit/utils.test.ts:40]: this looks wrong.
If I do `jointPaths("/tmp", "foo.txt")` I expect `/tmp/foo.txt`, not `tmp/foo.txt`

**Review by antocuni — APPROVED** [review]
good job :muscle: 
I'm approving it in advance so that you can merge it without any further review from me, but the logic about "remove leading slashes" should be fixed before merging

File: pyscriptjs/tests/unit/utils.test.ts
Comment (madhur-tandon) [reply pyscriptjs/tests/unit/utils.test.ts:40]: Done, I only check for a leading slash on the first element though. Otherwise it will be very cumbersome.

File: pyscriptjs/src/utils.ts
Comment (madhur-tandon) [reply pyscriptjs/src/utils.ts]: Done in #927

**Pull request merged**
